platform/stm32f0/drivers/componolit-runtime-drivers-gpio.ads

```diff
@@ -76,4 +92,36 @@ private
       Size => 32,
       Pack;
 
+   pragma Warnings (Off, "* bits of * unused");
+   type Bank_Config is record
+      Port_Mode : Mode_Register;
+      Pull_Mode : Pull_Register;
+   end record with
+      Size => 1024 * 8;
+
+   for Bank_Config use record
+      Port_Mode at 0 range 0 .. 31;
+      Pull_Mode at 12 range 0 .. 31;
+   end record;
```

Comment:
Are you sure that such holes are safe?

Reply:
How else would I annotate these holes? I mean that's what the representation clause does, defining where an entry begins and how long it is.

Reply:
Sure, but you also had to suppress an unused bits warning, which I guess exists for a reason. I was just wondering what the semantics of those holes is - can the compiler generated loads/stores for those regions to optimize the code (because they are just holes)? If so, this could have an effect on registers we are simply not modelling.  Maybe *reserved* elements are required. We should have a look into the Ada RM to find out what we can assumed about those undefined bit ranges.

Reply:
> Maybe _reserved_ elements are required. We should have a look into the Ada RM to find out what we can assumed about those undefined bit ranges.

The Ada RM says nothing about holes in representation. As far as the language is concerned, these memory locations simply don't exist.

Of course, in real implementations it's a bit more [complicated](https://docs.adacore.com/gnat_rm-docs/html/gnat_rm/gnat_rm/representation_clauses_and_pragmas.html#handling-of-records-with-holes)

General rule of thumb here is that the compiler tries it's best to avoid touching these holes at all (which in this context would generally mean, it will issue 32-bit instructions, even if more efficient 64-bit instruction may be available), but sometimes it's just plain impossible (for example if such holes are within a single byte). In that case the compiler will generate load/mask/store instructions to not touch these bits. Which, of course, in the context of a memory mapped register may still have unintended side effects.

Reply:
> for example if such holes are within a single byte

This is not the case here since both the holes and the blocks in the record are 32bit aligned and multiples of 32bit in size.

> but you also had to suppress an unused bits warning

This warning is in general issued when the size of the components of a record does not equal the size annotated with the `Size` aspect. This also happens if there are no holes or not even a representation clause.

Reply:
Thanks @Jellix for the pointer (the section is even named "Handling of Records with Holes" ;)

IMHO it should be safe for our 32bit sized and 32bit aligned values. We should just be cautious in situations where this is not the case.